haystack/generator/transformers.py

```diff
@@ -18,91 +19,227 @@ class RAGGeneratorType(Enum):
+        contextualized_embedding = self.tokenizer.question_encoder.batch_encode_plus(
             rag_input_strings,
-            max_length=300,
+            max_length=self.model.config.max_length,
             return_tensors=return_tensors,
             padding="max_length",
             truncation=True,
         )
 
-        return contextualized_inputs["input_ids"], contextualized_inputs["attention_mask"]
+        return contextualized_embedding["input_ids"].to(self.device), \
+               contextualized_embedding["attention_mask"].to(self.device)
+
+    def _embed_passage(self, title: str, text: str) -> torch.Tensor:
+        """Compute the DPR embeddings of document passages"""
+        input_ids = self.passage_tokenizer(
+            title,
+            text,
+            truncation=True,
+            padding="longest",
+            return_tensors="pt"
+        )["input_ids"]
+
+        embeddings = self.passage_encoder(
+            input_ids.to(device=self.device),
+            return_dict=True
+        ).pooler_output
+
+        return embeddings.detach().to(self.device)
+
+    def embed_passage_in_tensor(self, titles: List[str], texts: List[str],
+                                embeddings: List[Optional[numpy.ndarray]]) -> torch.Tensor:
+
+        embeddings_in_tensor: List[torch.Tensor] = []
+
+        # If of document missing embedding, then need embedding for all the documents
+        is_embedding_required = embeddings.__contains__(None)
+
+        if is_embedding_required:
+            for title, text in zip(titles, texts):
+                embeddings_in_tensor.append(
+                    self._embed_passage(
+                        title=title,
+                        text=text
+                    )
+                )
+        else:
+            embeddings_in_tensor = [torch.from_numpy(embedding) for embedding in embeddings]
+
+        return torch.cat(embeddings_in_tensor, dim=0).to(self.device)
 
     def predict(self, question: str, documents: List[Document], top_k: Optional[int] = None):
-        if len(documents) == 0 or documents[0].embedding is None:
-            raise AttributeError("generator need documents with embedding")
+        if len(documents) == 0:
+            raise AttributeError("generator need documents to predict the answer")
+
+        top_k_answers = top_k if top_k is None else self.top_k_answers
+
+        # Flatten the documents so easy to reference
+        flat_docs_dict = {}
+        for document in documents:
+            for k, v in document.__dict__.items():
+                if k not in flat_docs_dict:
+                    flat_docs_dict[k] = []
+                flat_docs_dict[k].append(v)
+
+        # Extract title
+        titles = [d.meta["name"] if d.meta and "name" in d.meta else "" for d in documents]
+
+        # Raw document embedding and set device of question_embedding
+        passage_embedding = self.embed_passage_in_tensor(
```

Comment:
In terms of variable naming we should be careful with singular / plural to avoid confusion. I believe here it's actually `passage_embeddings`